src/Database/LSMTree/Internal/BloomFilterQuery1.hs

```diff
@@ -84,47 +89,57 @@ bloomQueries ::
   where
-    !rsN = V.length blooms
-    !ksN = V.length ks
-
-    hs :: VP.Vector (Bloom.Hashes SerialisedKey)
-    !hs  = VP.generate ksN $ \i -> Bloom.hashes (V.unsafeIndex ks i)
+    !rsN = V.length filters
+    !ksN = V.length keys
+    !keyhashes = P.generatePrimArray (V.length keys) $ \i ->
+                   Bloom.hashes (V.unsafeIndex keys i)
 
-    -- Loop over all run indexes
+    -- loop over all filters
     loop1 ::
-         VPM.MVector s RunIxKeyIx
+         P.MutablePrimArray s RunIxKeyIx
       -> ResIx
       -> RunIx
-      -> ST s (VPM.MVector s RunIxKeyIx)
-    loop1 !res1 !resix1 !rix
-      | rix == rsN = pure $ VPM.slice 0 resix1 res1
-      | otherwise
-      = do
-          (res1', resix1') <- loop2 res1 resix1 0 (blooms `V.unsafeIndex` rix)
-          loop1 res1' resix1' (rix+1)
+      -> ST s (P.MutablePrimArray s RunIxKeyIx)
+    loop1 !res !resix !rix | rix == rsN = P.resizeMutablePrimArray res resix
+    loop1 !res !resix !rix = do
+        loop2_prefetch 0
+        (res', resix') <- loop2 res resix 0
+        loop1 res' resix' (rix+1)
       where
-        -- Loop over all key indexes
+        !filter = V.unsafeIndex filters rix
+
+        -- loop over all keys
         loop2 ::
-             VPM.MVector s RunIxKeyIx
+             P.MutablePrimArray s RunIxKeyIx
           -> ResIx
           -> KeyIx
-          -> Bloom SerialisedKey
-          -> ST s (VPM.MVector s RunIxKeyIx, ResIx)
-        loop2 !res2 !resix2 !kix !b
+          -> ST s (P.MutablePrimArray s RunIxKeyIx, ResIx)
+        loop2 !res2 !resix2 !kix
           | kix == ksN = pure (res2, resix2)
-          | let !h = hs `VP.unsafeIndex` kix
-          , Bloom.elemHashes b h = do
-              -- Double the vector if we've reached the end.
-              -- Note unsafeGrow takes the number to grow by, not the new size.
-              res2' <- if resix2 == VPM.length res2
-                        then VPM.unsafeGrow res2 (VPM.length res2)
-                        else pure res2
-              VPM.unsafeWrite res2' resix2 (RunIxKeyIx rix kix)
-              loop2 res2' (resix2+1) (kix+1) b
-          | otherwise = loop2 res2 resix2 (kix+1) b
+          | let !keyhash = P.indexPrimArray keyhashes kix
+          , Bloom.elemHashes filter keyhash = do
+              P.writePrimArray res2 resix2 (RunIxKeyIx rix kix)
+              ressz2 <- P.getSizeofMutablePrimArray res2
+              res2'  <- if resix2+1 < ressz2
+                         then return res2
+                         else P.resizeMutablePrimArray res2 (ressz2 * 2)
+              loop2 res2' (resix2+1) (kix+1)
+
+          | otherwise =
+              loop2 res2 resix2 (kix+1)
+
+        loop2_prefetch :: KeyIx -> ST s ()
+        loop2_prefetch !kix
+          | kix == ksN = pure ()
+          | otherwise  = do
+              let !keyhash = P.indexPrimArray keyhashes kix
+              Bloom.prefetchElem filter keyhash
+              loop2_prefetch (kix+1)
```

Comment:
The commit message for 8eaf7cef9b5ad894a8aa981c6440d323ff8a3d3c says that the prefetch distance is the number of runs, but this looks like the prefetch distance is the number of keys

Reply:
Yes, oops. I kept changing my mind. I did both versions and benchmarked them.

Reply:
Updated.

So in the benchmarks I did, nesting the loops this way round was slightly faster. But arguably, the other way round should have more stable prefetching behaviour, since it would depend on number of runs and not the lookup batch size.

Switching it round is not that hard. I could still do that. Thoughts?

Reply:
I've switched it round.